### PR TITLE
Use BSER to communicate with watchman.

### PR DESCRIPTION
**Pull request by mcguigan**

### CODE REVIEW

**Comment by dturner-tw** [comment]
The second commit is causing a build failure in CI.

File: bser_parse.c [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+static bser_t* new_compact_object(
+    bser_buffer_t* buffer, bser_t* header, bser_t* fill)
+{
+    assert(bser_is_array(header));
+    size_t sz = bser_array_size(header);
+    bser_key_value_pair_t* fields = malloc(sizeof(*fields) * sz);
+    if (fields == NULL) {
+        new_error("Could not allocate enough memory to hold "
```
Comment (dturner-tw) [inline bser_parse.c]: should be return new_error?

File: bser_parse.c [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_UNPARSED;
+    fill->value.unparsed.buffer = buffer;
+    return fill;
+}
+
+static bser_t* new_error(const char* message, bser_t* fill)
```
Comment (dturner-tw) [inline bser_parse.c]: __attribute__((warn_unused_result)) ?

File: bser_parse.c (hunk 1 above)
Comment (dturner-tw) [inline bser_parse.c]: + more instances below

File: watchman.c [hunk 3]
```diff
@@ -224,34 +234,40 @@ static json_t *read_json_with_timeout(int fd, char* buf, size_t bytes, struct ti
...
-        if (json)
-            return json;
+        proto_t proto;
+        if (use_bser_encoding) {
+            bser_t* bser = bser_parse_buffer((uint8_t*)buf, read_so_far, NULL);
+            proto = proto_from_bser(bser);
+        } else {
+            json_error_t jerror;
+            json_t* json = json_loads(buf, JSON_DISABLE_EOF_CHECK, &jerror);
+                        proto = proto_from_json(json);
```
Comment (dturner-tw) [inline watchman.c]: nit: indentation

File: bser_parse.c (hunk 1 above)
Comment (mcguigan) [inline bser_parse.c]: Yes, that should have been a return new_error().  But instead I changed it to a void method (the return value was unused in the only place it was called).  new_error() fills in the passed struct, so the other uses are correct.

File: bser_parse.c (hunk 2 above)
Comment (mcguigan) [inline bser_parse.c]: Nah, in many cases it's perfectly fine to use new_error (or new_unparsed) only for its side-effect of filling in the 'fill' struct, but ignoring the return value.

File: bser_parse.c (hunk 1 above)
Comment (dturner-tw) [inline bser_parse.c]: But you still need to return here because otherwise you're going to write to a NULL pointer.

Comment (mcguigan) [inline bser_parse.c]: I updated the commit, but it seems to still displaying the old content here.  Take a look at this function (new_compact_object) here: https://github.com/twitter/libwatchman/pull/16/commits/0671f0ff8b568fede7733299fea0ff2439da6263

Comment (mcguigan) [inline bser_parse.c]: Oh, nevermind, I see what you were saying now.  I'll fix this.

File: bser.c [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+const char* bser_error_message(bser_t* bser)
+{
+    assert(bser_is_error(bser));
+    return bser->value.error_message;
+}
+
+bser_t*
+bser_new_integer(uint64_t value, bser_t* fill)
```
Comment (dturner-tw) [inline bser.c]: should be int64_t not uint64_t.

File: bser_parse.c [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+bser_t*
+bser_parse_content(uint8_t* data, size_t buflen, bser_t* fill)
+{
+    bser_buffer_t* read_buffer = new_read_buffer(data, buflen, 0);
+		/* Create lazily-parsed node that will self-parse when it is accessed */
+    return new_unparsed(read_buffer, fill);
+}
+
+static int
+integer_from_file(FILE* fp, uint64_t* value)
```
Comment (dturner-tw) [inline bser_parse.c]: Should be int64_t?

File: bser.c (hunk 4 above)
Comment (mcguigan) [inline bser.c]: Yes, will fix.

File: bser_parse.c (hunk 5 above)
Comment (mcguigan) [inline bser_parse.c]: Yes, and I added a clause to check for a non-positive size, too.

File: bser_private.h [hunk 6]
```diff
@@ -53,7 +49,7 @@ typedef struct bser {
             struct bser_key_value_pair* fields;
             size_t length;
         } object;
-        struct bser_unparsed unparsed;
+				struct bser_buffer* unparsed;
```
Comment (dturner-tw) [inline bser_private.h]: ?

File: proto.h [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+void
+proto_decref(proto_t p)
+{
+    if (p.type == PROTO_JSON) {
+        json_decref(p.u.json);
+        p.u.json = NULL;
+    } else {
+        bser_free(p.u.bser);
```
Comment (dturner-tw) [inline proto.h]: Is it really safe to free here?  That is, do we only ever have one reference to an object? It's probably worth at least calling this out in a comment.

File: bser.c [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    bser_free_contents(bser);
+    free(bser);
+}
+
+int
+bser_string_strcmp(const char* match, bser_t* bser)
+{
+    assert(bser_is_string(bser));
+    bser_parse_if_necessary(bser);
+    int cmp = memcmp(match, bser->value.string.chars, bser->value.string.length);
```
Comment (dturner-tw) [inline bser.c]: wait, what if match is shorter than bser->value.string.length?

File: bser.c [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+        }
+    }
+    return NULL;
+}
+
+bser_t*
+bser_array_get(bser_t* bser, size_t index)
+{
+    bser_parse_if_necessary(bser);
+    assert(bser_is_array(bser));
```
Comment (dturner-tw) [inline bser.c]: assert index < bser->value.array.length ?

File: bser_parse.c [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+    } else if (bser_is_array(bser) && (index = bser_array_size(bser)) > 0) {
+        return is_fully_parsed(&bser->value.array.elements[index - 1]);
+    } else if (bser_is_object(bser) && (index = bser_object_size(bser)) > 0) {
+        return is_fully_parsed(&bser->value.object.fields[index - 1].value);
+    }
+    return 1;
+}
+
+static void
+fully_parse(bser_t* bser) {
```
Comment (dturner-tw) [inline bser_parse.c]: nit: newline before {

File: tests/t/int_small_signed.json [hunk 11]
```diff
@@ -0,0 +1 @@
+[ 253 ]
```
Comment (dturner-tw) [inline tests/t/int_small_signed.json]: "signed" is a funny name for this value.  I mean, yes, it has some sign, but...

**Comment by dturner-tw** [comment]
OK, review completed -- please fix (or tell me why I'm wrong about) the remaining errors & we'll be ready to merge.

File: proto.h (hunk 7 above)
Comment (mcguigan) [inline proto.h]: I'll change the name to 'proto_free()' and add a note that the json is only decref'd.  In all the places we use this in watchman.c, there's only one ref to the object.

File: bser.c (hunk 8 above)
Comment (mcguigan) [inline bser.c]: Well, memcmp will return the right value, as there will be a \0 in position match[strlen(match)].  However, memcmp() may read beyond the end of the string.  I'll address this.

File: bser.c (hunk 9 above)
Comment (mcguigan) [inline bser.c]: Yep.

File: tests/t/int_small_signed.json (hunk 11 above)
Comment (mcguigan) [inline tests/t/int_small_signed.json]: I'll change it to int_small_signbit, as it's an edge case test on bit 0x80.

**Comment by mcguigan** [comment]
Ok, all fixes are in and this is ready for a pull.

**Comment by dturner-tw** [comment]
Is github lying to me, or is 33ef338 the latest version on this PR?  That version doesn't seem to have fixes for my comments?

**Comment by mcguigan** [comment]
yeah that's the latest, but maybe I messed up dropped some of the changes
when screwing around with branches.  Tag what is missing and I'll go back
and find or fix those.

On Fri, Apr 22, 2016 at 2:32 PM, David Turner <notifications@github.com>
wrote:

> Is github lying to me, or is 33ef338
> <https://github.com/twitter/libwatchman/commit/33ef338c0ff575b978852f573643f53d3f4318e4>
> the latest version on this PR? That version doesn't seem to have fixes for
> my comments?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/libwatchman/pull/16#issuecomment-213544966>
>



-- 

[image: twitter-icon-large.png]

Keith McGuigan

@kamggg

kmcguigan@twitter.com
